examples/nlp/mlm_training_tpus.py

```diff
@@ -0,0 +1,469 @@
+    if shuffle:
+        assert shuffle_buffer_size is not None
+        dataset = dataset.shuffle(shuffle_buffer_size)
+    dataset = dataset.batch(batch_size, drop_remainder=True)
+    dataset = dataset.map(mask_fn, num_parallel_calls=auto)
+    dataset = dataset.prefetch(auto)
+    return dataset
+
+
+"""
+Let's prepare our datasets with these utilities: 
+"""
+
+per_replica_batch_size = 16  # Change as needed.
+batch_size = per_replica_batch_size * strategy.num_replicas_in_sync
+shuffle_buffer_size = 2**18  # Default corresponds to a 1GB buffer for seq_len 512
+
+train_dataset = prepare_dataset(
+    training_records,
+    decode_fn=decode_fn,
+    mask_fn=mask_with_collator,
+    batch_size=batch_size,
+    shuffle=True,
+    shuffle_buffer_size=shuffle_buffer_size,
+)
+
+eval_dataset = prepare_dataset(
+    eval_records,
+    decode_fn=decode_fn,
+    mask_fn=mask_with_collator,
+    batch_size=batch_size,
+    shuffle=False,
+)
+
+"""
+Let's now investigate how a single batch of dataset looks like. 
+"""
+
+single_batch = next(iter(train_dataset))
+print(single_batch.keys())
+
+"""
+* `input_ids` denotes the tokenized versions of the input samples containing the mask
+tokens as well. 
+* `attention_mask` denotes the mask to be used when performing attention operation.s 
+* `labels` denotes the actual values of masked tokens the model is supposed to learn from.
+"""
+
+for k in single_batch:
+    if k == "input_ids":
+        input_ids = single_batch[k]
+        print(f"Input shape: {input_ids.shape}")
+    if k == "labels":
+        labels = single_batch[k]
+        print(f"Label shape: {labels.shape}")
+
+"""
+Now, we can leverage our `tokenizer` to investigate the values of the tokens. Let's start
+with `input_ids`: 
+"""
+
+idx = 0
+print("Taking the first sample:\n")
+print(tokenizer.decode(input_ids[idx].numpy()))
+
+"""
+As expected, the decoded tokens contain the special tokens including the mask tokens as
+well. Let's now investigate the mask tokens: 
+"""
+
+# Taking the first 30 tokens of the first sequence.
+print(labels[0].numpy()[:30])
+
+"""
+Here, `-100` means that the corresponding tokens in the `input_ids` are NOT masked and
+non `-100` values denote the actual values of the masked tokens. 
+"""
+
+"""
+## Initialize the mode and and the optimizer
+"""
+
+"""
+With the datasets prepared, we now initialize and compile our model and optimizer within
+the `strategy.scope()`:
+"""
+
+# For this example, we keep this value to 10. But for a realistic run, start with 500.
+num_epochs = 10
+steps_per_epoch = num_train_samples // (
+    per_replica_batch_size * strategy.num_replicas_in_sync
+)
+total_train_steps = steps_per_epoch * num_epochs
+learning_rate = 0.0001
+weight_decay_rate = 1e-3
+
+with strategy.scope():
+    model = TFAutoModelForMaskedLM.from_config(config)
+    model(
+        model.dummy_inputs
+    )  # Pass some dummy inputs through the model to ensure all the weights are built
```

Comment:
This is not very elegant -- canonically you'd use the `build(input_shape)` method to build the model's state. In this case I suppose you could implement it as:

```
def build(self, input_shape=None):
    self(self.dummy_inputs)
    self.built = True
```

Then you can just do: `model.build()`.

Reply:
I will leave this for @RocketKnight1 to comment further.

Reply:
> I've been contemplating a community push to add proper build() methods everywhere, but every layer would need its own ones and it's quite a lot of code.

Why not just have one `build()` on the top-level model, implemented using your dummy inputs system?

> Also, do we need to set self.built after self(self.dummy_inputs)? I thought that build() would be called on the model and on each layer when the network is first called, which would set self.built for us.

Actually yes, it will already be set after `__call__()`.

Reply:
Ah, I didn't even think of that! I assumed it was somehow illegal to pass inputs through the model in the `build()` method! That's actually quite helpful for getting us more in-line with Keras best practices - I'll experiment with it and possibly file a PR soon.

Reply:
>  I assumed it was somehow illegal to pass inputs through the model in the build() method! 

Ideally `build()` should be zero-flops -- so you can pass symbolic inputs, but hopefully not *eager* inputs.

You're allowed to do whatever you want in `build()`, there's just one rule: the only side effects should be building the model's variables. There should be no other side effect, ideally (e.g. no change to the values of existing non-trainable variables like BN stats, etc.).